controlplane/rosa/controllers/rosacontrolplane_controller.go

```diff
@@ -0,0 +1,445 @@
+		return fmt.Errorf("failed adding a watch for ROSACluster")
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;delete;patch
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinepools,verbs=get;list;watch
+
+// Reconcile will reconcile RosaControlPlane Resources.
+func (r *ROSAControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := logger.FromContext(ctx)
+
+	// Get the control plane instance
+	rosaControlPlane := &rosacontrolplanev1.ROSAControlPlane{}
+	if err := r.Client.Get(ctx, req.NamespacedName, rosaControlPlane); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Get the cluster
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, rosaControlPlane.ObjectMeta)
+	if err != nil {
+		log.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	if capiannotations.IsPaused(cluster, rosaControlPlane) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	rosaScope, err := scope.NewROSAControlPlaneScope(scope.ROSAControlPlaneScopeParams{
+		Client:         r.Client,
+		Cluster:        cluster,
+		ControlPlane:   rosaControlPlane,
+		ControllerName: strings.ToLower(rosaControlPlaneKind),
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	// Always close the scope
+	defer func() {
+		if err := rosaScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	if !rosaControlPlane.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Handle deletion reconciliation loop.
+		return r.reconcileDelete(ctx, rosaScope)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcileNormal(ctx, rosaScope)
+}
+
+func (r *ROSAControlPlaneReconciler) reconcileNormal(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (res ctrl.Result, reterr error) {
+	rosaScope.Info("Reconciling ROSAControlPlane")
+
+	// if !rosaScope.Cluster.Status.InfrastructureReady {
+	//	rosaScope.Info("Cluster infrastructure is not ready yet")
+	//	return ctrl.Result{RequeueAfter: r.WaitInfraPeriod}, nil
+	//}
+	if controllerutil.AddFinalizer(rosaScope.ControlPlane, ROSAControlPlaneFinalizer) {
+		if err := rosaScope.PatchObject(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Create the cluster:
+	clusterBuilder := cmv1.NewCluster().
```

Comment:
I'm guessing these we'll be able to configure eventually? All the hardcoded things like `ID("rosa")` or the FIPS(false) etc...

Reply:
yes, most of the fields in the code here would be configurable. This is just an early POC and we want it to be as minimal as possible until we flesh out the API.